### ISSUE TITLE
Configure a behaviour test runner

### ISSUE THREAD

**Issue by jGleitz** [HelfenKannJeder/come2help-web#26]
In the moment, we can run tests in one browser at a time. The goal would be to run browserstack tests in browsers specified in a file.

Possible solutions:
- multiple Travis builds (would make stuff clear but have the install overhead)
- self-written test run script (not a big deal actually)
- using [karma](karma-runner.github.io/) (I don't quite see how to use selenium with karma, but there is [karma-browserstack-launcher](https://github.com/karma-runner/karma-browserstack-launcher). I don't quite grasp how it works, though)

**Comment by vzickner** [comment]
Multiple travis builds would be executed parallel. This is maybe not allowed/wanted from BrowserStack?

**Comment by jGleitz** [comment]
a) We could maybe ask browserstack for an upgrade.
b) Travis offers an option to limit the number of concurrent builds.

**Comment by jGleitz** [comment]
Having read more about karma, I guess it's great for unit tests, but not really suitable for behaviour tests.

I think we should just write our own runner.

**Comment by BassT** [comment]
Did anybody consider https://github.com/browserstack/browserstack-runner?

**Comment by jGleitz** [comment]
Yes, and I like it's way of specifying browsers.

But as far as I understand, it does not give us an instance of a selenium webdriver.

**Comment by jGleitz** [comment]
I'd appreciate if you'd have another look at it, but as far as I see, [browserstack-runner](https://github.com/browserstack/browserstack-runner) offers roughly the same functionality as [karma-browserstack-launcher](https://github.com/karma-runner/karma-browserstack-launcher) and is suited for _unit testing_. But as we plan to use karma anyway, I'd rather use the latter.

For behaviour tests, I like the idea of using selenium, and for that we need to run the tests multiple times, only with different instances of a selenium webdriver for different browserstack browsers. I did not find a tool that does that.

**Comment by BassT** [comment]
Alright, I see what you're saying now. So you're suggesting to write our own script for integration testing similar to [this](http://red-badger.com/blog/2014/05/27/automated-cross-browser-testing-with-browserstack-and-circleci/).

I thinks that's a good idea.

**Comment by jGleitz** [comment]
Updates on this:
- I wrote a quite well functioning test runner. 
- After I did that, I had another look at [Protractor](https://angular.github.io/protractor/#/) and realised that it would make quite sense to use it. (Apart from spoling the browserstack logs, it's a nice tool). 
- I started modifying the code to run protractor, which wasn't much work.
- But protractor seems not to work on iOs devices on browserstack.

That's were I'm stuck in the moment.

**Comment by BassT** [comment]
Did the test runner you implemented work fine? Why not just use that one? Any advantages in using Protractor?

**Comment by jGleitz** [comment]
The one I wrote works well. 

Protractor knows about Angular and automatically waits for its actions. So no need for waits anymore. E.g. if you hit a button that triggers some API calls, Protractor will wait for HTTP communication and controller action to finish. 
This is actually pretty cool, as writing error prone waits is all but trivial, as I already noticed in the small example tests. 

Plus, Protractor offers ways to access the angular parts of the DOM, which further abstracts from the actual HTML -  another good thing. 

On the other hand, in my recent tests I noticed that Protractor also has issues with Microsoft Edge and some older (pre 5) Android Phones. 
I have to do some more testing on my runner to be sure it does not have such problems, though.

**Comment by jGleitz** [comment]
More updates:

I did more tests the last days and am now very sure that we really should use protractor – it offers a lot of useful features. I'm in touch with the browserstack team to sort things out and hopefully we'll get the tests running on all relevant devices soon.

I wrote a lot of code for the standalone runner and modified that to support protractor. I now figured that this approach is suboptimal in many ways and the code actually belongs into protractor itself. 

To sum things up, this issue will take a little longer until implemented. I'll maybe make a PR soon that allows behaviour testing at least in some browsers to have something to start with.

**Comment by vzickner** [comment]
implemented in #31
